### PR TITLE
KT-52141: Optimize Java class snapshotting with ASM ClassWriter

### PR DESCRIPTION
**Pull request by hungvietnguyen**

### CODE REVIEW

File: build-common/src/org/jetbrains/kotlin/incremental/IncrementalJvmCache.kt [hunk 1]
```diff
@@ -757,33 +757,35 @@ private class ConstantsClassVisitor : ClassVisitor(Opcodes.API_VERSION) {
...
 
-        return object : MethodVisitor(Opcodes.API_VERSION, dummyClassWriter.visitMethod(0, name, desc, null, exceptions)) {
-            override fun visitEnd() {
-                val jvmName = name + desc
-                if (jvmName !in inlineFunctionNames) return
+        val classWriter = ClassWriter(0)
 
-                val dummyBytes = dummyClassWriter.toByteArray()!!
+        // The `version` and `name` parameters are important (see KT-38857), the others can be null.
+        classWriter.visit(/* version */ classVersion!!, /* access */ 0, /* name */ "ClassWithOneMethod", null, null, null)
```
Comment (gavra0) [inline build-common/src/org/jetbrains/kotlin/incremental/IncrementalJvmCache.kt:784]: Why did you remove "Object" as supertype?

File: compiler/incremental-compilation-impl/src/org/jetbrains/kotlin/incremental/classpathDiff/JavaClassSnapshotter.kt [hunk 2]
```diff
@@ -51,50 +47,49 @@ object JavaClassSnapshotter {
...
-        } else {
-            val abiValue = gson.toJson(javaElement)
-            val abiHash = abiValue.toByteArray().md5()
-            JavaElementSnapshot(javaElementName, abiHash)
-        }
+    private fun emptyClass() = ClassNode().also {
+        // We need to provide some minimal info to the class:
+        //   - Name is required.
+        //   - Class version is required if method bodies are considered, but we have removed method bodies in this class, so it's optional.
+        //   - Other info is optional.
```
Comment (gavra0) [inline compiler/incremental-compilation-impl/src/org/jetbrains/kotlin/incremental/classpathDiff/JavaClassSnapshotter.kt:92]: Does ASM complain about other flags e.g if you add an abstract method to a non-abstract class? One option that we discussed today is to copy all params from the original class.

File: build-common/src/org/jetbrains/kotlin/incremental/IncrementalJvmCache.kt (hunk 1 above)
Comment (hungvietnguyen) [reply build-common/src/org/jetbrains/kotlin/incremental/IncrementalJvmCache.kt:784]: Nice catch, the `ClassNode()` approach in `JavaClassSnapshotter` has supertype = `null` so I do the same here for consistency.

File: compiler/incremental-compilation-impl/src/org/jetbrains/kotlin/incremental/classpathDiff/JavaClassSnapshotter.kt (hunk 2 above)
Comment (hungvietnguyen) [reply compiler/incremental-compilation-impl/src/org/jetbrains/kotlin/incremental/classpathDiff/JavaClassSnapshotter.kt:92]: No, it doesn't complain.

I think there are a few levels of info:
  1. Name + class version only
  2. Name + class version + superName
  3. Properties provided by `ClassVisitor#visit`: name + class version + access + signature + superName + interfaces
  4. All possible class properties: The above + attrs + innerClasses + invisibleAnnotations + invisibleTypeAnnotations + module + nestHostClass + nestMembers + ...
  
I chose 1 because it's minimal and I've run this on all 600 jars of the Kotlin repo without a problem (we also have integration tests as a safety net).
 
Do you suggest a different level? (For consistency, I probably will apply the same level to both the "ClassNode" approach here and the "ClassVisitor" approach in `IncrementalJvmCache`.)

File: build-common/src/org/jetbrains/kotlin/incremental/IncrementalJvmCache.kt (hunk 1 above)
Comment (gavra0) [reply build-common/src/org/jetbrains/kotlin/incremental/IncrementalJvmCache.kt:784]: As long as ASM does not fail and it generates class that we can hash, it LGTM.

File: compiler/incremental-compilation-impl/src/org/jetbrains/kotlin/incremental/classpathDiff/JavaClassSnapshotter.kt (hunk 2 above)
Comment (gavra0) [reply compiler/incremental-compilation-impl/src/org/jetbrains/kotlin/incremental/classpathDiff/JavaClassSnapshotter.kt:92]: kk, that sounds like a good check, we can keep this as-is.

**Review by gavra0 — APPROVED** [review]

**Review by Tapchicoma — APPROVED** [review]

**Comment by hungvietnguyen** [comment]
This PR has been merged: https://github.com/JetBrains/kotlin/commit/9eb3c7ed7624913b63ac17b7887af76c62a44a8d and https://github.com/JetBrains/kotlin/commit/a7403fc6157779117ca0264b540ad8cb3aa7a1a1. Thx!
